### ISSUE TITLE
[DISCUSSION] Exposing the underlying struct

### ISSUE THREAD

**Issue by bigs** [multiformats/go-multiaddr#100]
In my opinion, hiding the underlying multiaddr struct is a hinderance, particularly when working with serializations in golang (looking at flags and encoding/json). Is there any reason we couldn't simply expose the struct and keep its member hidden? This would enable us to implement interfaces that require zero values.

**Comment by bigs** [comment]
ping @Kubuxu @raulk

**Comment by lgierth** [comment]
Why not -- are you looking for more serializations than the existing MarshalBinary, MarshalText MarshalJSON?

**Comment by Stebalien** [comment]
So, this is an issue for _deserializing_ because we need a concrete struct to deserialize into. However, I'm not sure what the issue is for _serialization_.

Note: there's also the tricky fact that we now have _two_ concrete implementations: `Component` and `multiaddr`.

**Comment by raulk** [comment]
Yeah, I've encountered this hindrance in the past. I get the impression the goal was to enable diversity of multiaddr implementations, but I can't see a use case for that.

**Comment by Stebalien** [comment]
> Yeah, I've encountered this hindrance in the past. I get the impression the goal was to enable diversity of multiaddr implementations, but I can't see a use case for that.

IMO, we should implement `Multiaddr` for `net.TCPAddr`, etc. However, I'm not sure how well that would work with how we _currently_ use multiaddrs (it would probably lead to a bunch of allocations).

**Comment by raulk** [comment]
Ah, I see what you mean. Treating `Multiaddrs` as a canonical format with conversions from/to other formats. Unlike Rust, I don't think this is feasible in Go by making TCPAddr implement Multiaddr from outside the owning package. We could, of course, provide translation util functions, but that's what go-multiaddr-net does already, right?

**Comment by Stebalien** [comment]
Hm, no, you're right. We _could_ add newtypes but we can't implement it directly (I sometimes blissfully forget about go's type system).

**Comment by bigs** [comment]
i think if we keep the interface and just expose the strict as well, it should solve our problem

**Comment by Stebalien** [comment]
@bigs could you explain _how_ this solves the problem? Having to type assert to the concrete type is going to be annoying.

**Comment by bigs** [comment]
@Stebalien i misspoke earlier when i said serialization was the issue. it's only deserialization.

there's a somewhat prevalent pattern, for example in the datastores, of having something like

`NewBadgerDatastore(...) *BadgerDatastore`

where `* BadgerDatastore` is the receiver for methods implementing the `Datastore` interface. i think we could do that for Multiaddr. we could even expose the concrete type in another package if we wanted to have the name preserved.

being able to to say

```golang
var bytes []byte
var addr *multiaddr.Multiaddr
if err := addr.UnmarshalJSON(bytes); err != nil {
  panic(err)
}
```

makes it possible for users to seamlessly read in json objects with string representations of multiaddrs. we could keep the signature of `NewMultiaddr` the same if we so pleased. furthermore, any functions accepting the `Multiaddr` interface should have no problem.

**Comment by Stebalien** [comment]
So, the issue I have with that is that the struct would need to be:

```
type Thing struct {
  ID peer.ID
  Addrs []*ma.MultiaddrConcrete
}
```

But now, when I want to _create_ this thing, I have to convert from my `ma.Multiaddr` to a `*ma.MultiaddrConcrete`. Basically, this is going to suck either way: it's either going to be annoying when serializing/deserializing or at a higher layer when creating these structs.

I wonder if using refmt would help? refmt allows specifying explicit transforms so it may be possible to specify how to decode a multiaddr.

**Comment by bigs** [comment]
ah, you know, that's a fair point. i think it's still probably worth it, because it would only be *adding* functionality, not removing or hindering anything. that said, it would be wonderful to have some multiaddr magics for refmt! could be a convenience package.

**Comment by hsanjuan** [comment]
I am going to bump this. I have custom code to wrap multiaddresses into a type that knows how to deserialize itself. Having to do that is worse than anything.

Could I convince you to remove the interface(s) altogether and use a concrete struct type instead, since nowhere this is used as an interface with alternate implementations, and right now it is a pain? (This would be backwards compatible as this type would have all the methods from the interface).

**Comment by raulk** [comment]
@hsanjuan I'm convinced. Multiaddrs are used everywhere, this indirection has proven useless over time (we have zero alternative implementations), and multiaddrs are a _data containers_ per spec, not an abstraction over alternative data structures.

The above use case mentioned by @Stebalien -- using the Multiaddr as a canonical interface to things like TCPAddr -- is not feasible in go, because we can't extend types beyond their definition site. We are going to need translation/wrapper types/methods either way.

I'm very annoyed because the way it's implemented, we are totally unable to unmarshal multiaddrs, for obvious reasons. In fact, the unit tests blissfully use the private type. This is BOA (broken on arrival), in my opinion:

https://github.com/multiformats/go-multiaddr/blob/5b1de2f51ff2368d5ce94a659f15ef26be273cd0/multiaddr_test.go#L626-L629

Something basic I'd like to be able to do:

```go
	ma, err := multiaddr.NewMultiaddr("/ip4/127.0.0.1/tcp/5001")
	if err != nil {
		panic(err)
	}

	bytes, err := json.Marshal(ma)
	if err != nil {
		panic(err)
	}

	fmt.Println((string)(bytes))

	var addr multiaddr.Multiaddr
	err = json.Unmarshal(bytes, &addr)
	if err != nil {
		panic(err)
	}
	fmt.Println(addr)
```

```
panic: json: cannot unmarshal string into Go value of type multiaddr.Multiaddr
```

**Comment by Stebalien** [comment]
So, we _are_ now using multiple implementations. Specifically, we have `Component` and `multiaddr` to make parsing multiaddrs easier.

However, we _can_ just export `multiaddr` as `GenericMultiaddr` (or something like that). That would allow us to unmarshal as:

```go
type Bar struct {
	Addr ma.Multiaddr
}

func main() {
	b := Bar{Addr: new(ma.GenericMultiaddr)}
	json.Unmarshal([]byte(`{"Addr": "/ip4/1.2.3.4"}`), &b)
	fmt.Println(b)
}
```

But that's still kind of crappy as we need to partially initialize the struct.

---

An alternative is to use [refmt](https://github.com/polydawn/refmt) which allows specifying per-type decoders.

**Comment by jfietz** [comment]
As I just ran into this issue, has there been any further development on this? Specifically in the context of exchanging host information on testground without using a dht.

**Comment by mg98** [comment]
I will just make a dumb proposal: 

We could just refactor the [`Multiaddr`](https://github.com/multiformats/go-multiaddr/blob/master/interface.go) interface to something else, like `IMultiaddr` but less Java-flavored,
and then refactor the [`multiaddr`](https://github.com/multiformats/go-multiaddr/blob/master/multiaddr.go) struct to `Multiaddr`.
Furthermore, [`NewMultiaddr`](https://github.com/multiformats/go-multiaddr/blob/master/multiaddr.go#L17) should return `*Multiaddr` (the struct).

That's the most straightforward but also least-harmful way I can think of doing it. I will be happy to create a PR. Waiting for comments first, though.

**Comment by jcace** [comment]
Hey guys, ran into this error recently too while trying to unmarshall Multiaddrs from json. 
```cannot unmarshal string into Go struct field MinerChainInfo.chain_info.addresses of type multiaddr.Multiaddr```

Ended up having to redefine it as `string` and manually convert it over with `NewMultiAddr()` function. It would be nice if there was a more straightforward way of doing it 👍

**Comment by 2color** [comment]
Linking https://github.com/multiformats/go-multiaddr/pull/261 which refactors the whole library and exposes a concrete type instead of an interface.
